{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import itertools\n",
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from doit.tools import register_doit_as_IPython_magic\n",
    "register_doit_as_IPython_magic()\n",
    "\n",
    "from oudelaar_tiled_capture_2019 import config, doit_tasks\n",
    "datasets = config.datasets()\n",
    "doit_db = os.path.join(config.base_folder, '.doit.db')\n",
    "doit_opts = '--db-file {} --backend sqlite3'.format(doit_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bam_sample_name(bam_path):\n",
    "    return os.path.basename(os.path.dirname(bam_path))\n",
    "\n",
    "def bam_build(bam_path):\n",
    "    return os.path.basename(bam_path).split('.')[1]\n",
    "\n",
    "def swap_ext(input_filename, new_ext):\n",
    "    \"\"\"Remove the last extension from input_filename and replace it with new_ext\"\"\"\n",
    "    \n",
    "    root, ext = os.path.splitext(input_filename)\n",
    "    return root + new_ext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_peaks_table_path = config.in_data_processed('atac/fetal_liver_atac_counts.{}.table')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_peaks_table = {build: pd.read_csv(final_peaks_table_path.format(build), sep='\\t')\n",
    "                     for build in ('mm9', 'mm10')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "scaling_factors = {build: (1e6 / final_peaks_table[build].iloc[:,4:].sum())\n",
    "                     for build in ('mm9', 'mm10')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_file = {build: config.in_data_raw('annotation/{}.chrom.sizes'.format(build))\n",
    "                     for build in ('mm9', 'mm10')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_bams = glob.glob(\n",
    "    config.in_data_intermediate(\n",
    "        'atac/merged_technical_reps/*/filtered.*.sorted.bam'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedgraph_dir = config.in_data_intermediate('hub/atac')\n",
    "\n",
    "def bedgraph_path_from_bam(bam):\n",
    "    sample_name = bam_sample_name(bam)\n",
    "    build = bam_build(bam)\n",
    "    return os.path.join(bedgraph_dir, '{}.{}.bg'.format(sample_name, build))\n",
    "\n",
    "def bigwig_path_from_bam(bam):\n",
    "    return swap_ext(bedgraph_path_from_bam(bam), '.bw')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def task_generate_bedgraphs():\n",
    "    for bam in merged_bams:\n",
    "        sample_name = bam_sample_name(bam)\n",
    "        build = bam_build(bam)\n",
    "        scaling_factor = scaling_factors[build][sample_name]\n",
    "        build_genome = genome_file[build]\n",
    "        \n",
    "        yield {\n",
    "            'name': '{}-{}'.format(sample_name, build),\n",
    "            'file_dep': [bam],\n",
    "            'targets': [bedgraph_path_from_bam(bam)],\n",
    "            'actions': [\n",
    "                'bedtools genomecov -bg -scale {scaling_factor:.5f} '\n",
    "                '-g {genome_file} -ibam %(dependencies)s '\n",
    "                '| sort -k1,1 -k2,2n > %(targets)s'.format(\n",
    "                    scaling_factor=scaling_factor,\n",
    "                    genome_file=build_genome),\n",
    "            ]\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_details_from_path(filepath):\n",
    "    fname = os.path.basename(filepath)\n",
    "    sample_id, build, ext = fname.split('.')\n",
    "    parts = sample_id.split('_')\n",
    "    library_ids = parts.pop()\n",
    "    expt_ids = parts.pop()\n",
    "    sample_name = '_'.join(parts)\n",
    "    return sample_name, expt_ids, library_ids, build\n",
    "    \n",
    "\n",
    "def linked_bw_filename(bw_path):\n",
    "    sample_name, expt_ids, library_ids, build = sample_details_from_path(bw_path)\n",
    "    return '{}_{}.{}.bw'.format(sample_name, expt_ids, build)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_merged_exptid(expt_list):\n",
    "    if len(expt_list) == 1:\n",
    "        return expt_list[0]\n",
    "    else:\n",
    "        expt_nos = sorted([int(e[4:]) for e in expt_list])\n",
    "\n",
    "        return 'expts_{}'.format('-'.join([str(s) for s in expt_nos]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def task_generate_bigwigs():\n",
    "\n",
    "    bedgraphs_to_merge = {\n",
    "        'mm9': defaultdict(list),\n",
    "        'mm10': defaultdict(list)\n",
    "    }\n",
    "    \n",
    "    for t in task_generate_bedgraphs():\n",
    "\n",
    "        bedgraph_path = t['targets'][0]\n",
    "        sample_name, expt_ids, library_ids, build = sample_details_from_path(bedgraph_path)\n",
    "        bedgraphs_to_merge[build][sample_name].append(bedgraph_path)\n",
    "\n",
    "    for build, build_to_merge in bedgraphs_to_merge.items():\n",
    "        \n",
    "        build_genome = genome_file[build]\n",
    "        \n",
    "        for sample_name, sample_bedgraphs in build_to_merge.items():\n",
    "            \n",
    "            expt_ids = [sample_details_from_path(bg)[1] for bg in sample_bedgraphs]\n",
    "            \n",
    "            merged_expt_id = get_merged_exptid(expt_ids)\n",
    "            \n",
    "            target_bw = os.path.join(bedgraph_dir, '{}_{}.{}.bw'.format(\n",
    "                sample_name, merged_expt_id, build))\n",
    "            \n",
    "            if len(sample_bedgraphs) > 1:\n",
    "        \n",
    "                yield {\n",
    "                    'name': '{}-{}'.format(sample_name, build),\n",
    "                    'file_dep': sample_bedgraphs,\n",
    "                    'targets': [target_bw],\n",
    "                    'actions': [\n",
    "                        \"bedtools unionbedg -i %(dependencies)s | \"\n",
    "                        \"awk '{sum = 0; for (i = 4; i <= NF; i++) sum += $i; sum /= (NF - 3); \"\n",
    "                        \"print $1\\\"\\t\\\" $2\\\"\\t\\\" $3\\\"\\t\\\" sum}' > %(targets)s.temp.bg\",\n",
    "\n",
    "                        'bedGraphToBigWig '\n",
    "                        '%(targets)s.temp.bg {genome_file} %(targets)s'.format(\n",
    "                            genome_file=build_genome),\n",
    "\n",
    "                        'rm %(targets)s.temp.bg'\n",
    "                    ]            \n",
    "                }\n",
    "            \n",
    "            else:\n",
    "                \n",
    "                yield {\n",
    "                    'name': '{}-{}'.format(sample_name, build),\n",
    "                    'file_dep': sample_bedgraphs,\n",
    "                    'targets': [target_bw],\n",
    "                    'actions': [\n",
    "                        'bedGraphToBigWig '\n",
    "                        '%(dependencies)s {genome_file} %(targets)s'.format(\n",
    "                            genome_file=build_genome)\n",
    "                    ]            \n",
    "                }\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def task_link_bigwigs():\n",
    "    for t in task_generate_bigwigs():\n",
    "        \n",
    "        final_bw = t['targets'][0]\n",
    "        bw_fname = os.path.basename(final_bw)\n",
    "        \n",
    "        link_path = os.path.join(\n",
    "            os.environ['PUBLIC_DIR'], 'bigwigs', bw_fname)\n",
    "        \n",
    "        yield {\n",
    "            'name': bw_fname,\n",
    "            'file_dep': [final_bw],\n",
    "            'targets': [link_path],\n",
    "            'actions': ['ln -s %(dependencies)s %(targets)s']\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "bigwig_cols = {\n",
    "    'HSPC': '127,205,187',\n",
    "    'LMPP': '65,182,196',\n",
    "    'MPP2': '29,145,192',\n",
    "    'MPP3': '34,94,168',\n",
    "    'Lin': '2,129,138',\n",
    "    'S0_CD71_low': '252,141,89',\n",
    "    'S0_CD71_medium': '252,141,89',\n",
    "    'S1': '239,101,72',\n",
    "    'S2': '215,48,31',\n",
    "    'S3': '179,0,0',\n",
    "}\n",
    "\n",
    "bigwig_text = \"\"\"\n",
    "    track {sample}\n",
    "    parent {tissue}-ATAC-seq on\n",
    "    shortLabel {sample}\n",
    "    longLabel {long_sample}\n",
    "    autoScale off\n",
    "    viewLimits 0:10\n",
    "    maxHeightPixels 100:100:8\n",
    "    type bigWig\n",
    "    bigDataUrl http://userweb.molbiol.ox.ac.uk{link_path}\n",
    "    color {color}\n",
    "        \n",
    "\"\"\"\n",
    "\n",
    "composite_text = \"\"\"\n",
    "track {tissue}-ATAC-seq\n",
    "compositeTrack on\n",
    "type bigWig\n",
    "shortLabel {tissue}-ATAC-seq\n",
    "longLabel Primary mouse {long_tissue} ATAC-seq\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "tissue_samples = {\n",
    "    'FL': ['Lin', 'S0_CD71_low', 'S0_CD71_medium', 'S1', 'S2', 'S3'],\n",
    "    'BM': ['HSPC', 'LMPP', 'MPP2', 'MPP3']\n",
    "}\n",
    "\n",
    "def make_bigwig_hub_file(dependencies, targets):\n",
    "    \n",
    "    output_file_path, = targets\n",
    "    \n",
    "    with open(output_file_path, 'w') as output_file:\n",
    "        \n",
    "        for tissue, long_tissue in ('FL', 'fetal liver'), ('BM', 'bone marrow'):\n",
    "            \n",
    "            print(composite_text.format(\n",
    "                        tissue=tissue, long_tissue=long_tissue),\n",
    "                  file=output_file)\n",
    "\n",
    "            for linked_bw in sorted(dependencies):\n",
    "\n",
    "                long_sample = os.path.basename(linked_bw).split('.')[0]\n",
    "\n",
    "                sample = long_sample.split('_expt')[0]\n",
    "                \n",
    "                if sample not in tissue_samples[tissue]:\n",
    "                    continue\n",
    "\n",
    "                print(\n",
    "                    bigwig_text.format(\n",
    "                        tissue=tissue,\n",
    "                        sample=sample,\n",
    "                        long_sample=long_sample,\n",
    "                        link_path=linked_bw,\n",
    "                        color=bigwig_cols[sample]), \n",
    "                    file=output_file)\n",
    "            \n",
    "def task_bigwig_hub_file():\n",
    "    \n",
    "    linked_bigwigs = [t['targets'][0] for t in task_link_bigwigs()]\n",
    "    \n",
    "    for build in 'mm9', 'mm10':\n",
    "        \n",
    "        build_bigwigs = [bw_path for bw_path in linked_bigwigs\n",
    "                         if build in bw_path]\n",
    "        \n",
    "        yield {\n",
    "            'name': 'Making hub ({})'.format(build),\n",
    "            'targets': [config.in_data_processed(\n",
    "                'hub/tracks_file/atac-bigwig.{build}.txt'.format(build=build))],\n",
    "            'file_dep': build_bigwigs,\n",
    "            'actions': [make_bigwig_hub_file],\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- generate_bedgraphs:S2_expt13_RBAT37-RBAT38-mm9\n",
      "-- generate_bedgraphs:S2_expt13_RBAT37-RBAT38-mm10\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt10_RBAT44-RBAT45-mm10\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt10_RBAT44-RBAT45-mm9\n",
      "-- generate_bedgraphs:LMPP_expt93_RBAT95-mm9\n",
      "-- generate_bedgraphs:LMPP_expt93_RBAT95-mm10\n",
      "-- generate_bedgraphs:S0_CD71_low_expt35_RBAT52-mm10\n",
      "-- generate_bedgraphs:S0_CD71_low_expt35_RBAT52-mm9\n",
      "-- generate_bedgraphs:S1_expt10_RBAT46-RBAT47-mm10\n",
      "-- generate_bedgraphs:S1_expt10_RBAT46-RBAT47-mm9\n",
      "-- generate_bedgraphs:Lin_expt35_RBAT61-RBAT62-mm10\n",
      "-- generate_bedgraphs:Lin_expt35_RBAT61-RBAT62-mm9\n",
      "-- generate_bedgraphs:HSPC_expt93_RBAT94-mm9\n",
      "-- generate_bedgraphs:HSPC_expt93_RBAT94-mm10\n",
      "-- generate_bedgraphs:S3_expt13_RBAT39-RBAT40-mm9\n",
      "-- generate_bedgraphs:S3_expt13_RBAT39-RBAT40-mm10\n",
      "-- generate_bedgraphs:MPP3_expt93_RBAT97-mm9\n",
      "-- generate_bedgraphs:MPP3_expt93_RBAT97-mm10\n",
      "-- generate_bedgraphs:S3_expt35_RBAT59-RBAT60-mm9\n",
      "-- generate_bedgraphs:S3_expt35_RBAT59-RBAT60-mm10\n",
      "-- generate_bedgraphs:S1_expt35_RBAT55-RBAT56-mm9\n",
      "-- generate_bedgraphs:S1_expt35_RBAT55-RBAT56-mm10\n",
      "-- generate_bedgraphs:S0_CD71_low_expt13_RBAT31-RBAT32-mm10\n",
      "-- generate_bedgraphs:S0_CD71_low_expt13_RBAT31-RBAT32-mm9\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt35_RBAT53-RBAT54-mm9\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt35_RBAT53-RBAT54-mm10\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt13_RBAT33-RBAT34-mm10\n",
      "-- generate_bedgraphs:S0_CD71_medium_expt13_RBAT33-RBAT34-mm9\n",
      "-- generate_bedgraphs:S2_expt35_RBAT57-RBAT58-mm9\n",
      "-- generate_bedgraphs:S2_expt35_RBAT57-RBAT58-mm10\n",
      "-- generate_bedgraphs:MPP2_expt93_RBAT96-mm10\n",
      "-- generate_bedgraphs:MPP2_expt93_RBAT96-mm9\n",
      "-- generate_bedgraphs:S0_CD71_low_expt10_RBAT42-RBAT43-mm9\n",
      "-- generate_bedgraphs:S0_CD71_low_expt10_RBAT42-RBAT43-mm10\n",
      "-- generate_bedgraphs:S1_expt13_RBAT35-RBAT36-mm9\n",
      "-- generate_bedgraphs:S1_expt13_RBAT35-RBAT36-mm10\n",
      "-- generate_bigwigs:S0_CD71_low-mm10\n",
      "-- generate_bigwigs:MPP2-mm10\n",
      "-- generate_bigwigs:HSPC-mm10\n",
      "-- generate_bigwigs:S2-mm10\n",
      "-- generate_bigwigs:S1-mm10\n",
      "-- generate_bigwigs:S0_CD71_medium-mm10\n",
      "-- generate_bigwigs:MPP3-mm10\n",
      "-- generate_bigwigs:LMPP-mm10\n",
      "-- generate_bigwigs:Lin-mm10\n",
      "-- generate_bigwigs:S3-mm10\n",
      "-- generate_bigwigs:S0_CD71_low-mm9\n",
      "-- generate_bigwigs:MPP2-mm9\n",
      "-- generate_bigwigs:HSPC-mm9\n",
      "-- generate_bigwigs:S2-mm9\n",
      "-- generate_bigwigs:S1-mm9\n",
      "-- generate_bigwigs:S0_CD71_medium-mm9\n",
      "-- generate_bigwigs:MPP3-mm9\n",
      "-- generate_bigwigs:LMPP-mm9\n",
      "-- generate_bigwigs:Lin-mm9\n",
      "-- generate_bigwigs:S3-mm9\n",
      "-- link_bigwigs:S0_CD71_low_expts_10-13-35.mm10.bw\n",
      "-- link_bigwigs:MPP2_expt93.mm10.bw\n",
      "-- link_bigwigs:HSPC_expt93.mm10.bw\n",
      "-- link_bigwigs:S2_expts_13-35.mm10.bw\n",
      "-- link_bigwigs:S1_expts_10-13-35.mm10.bw\n",
      "-- link_bigwigs:S0_CD71_medium_expts_10-13-35.mm10.bw\n",
      "-- link_bigwigs:MPP3_expt93.mm10.bw\n",
      "-- link_bigwigs:LMPP_expt93.mm10.bw\n",
      "-- link_bigwigs:Lin_expt35.mm10.bw\n",
      "-- link_bigwigs:S3_expts_13-35.mm10.bw\n",
      "-- link_bigwigs:S0_CD71_low_expts_10-13-35.mm9.bw\n",
      "-- link_bigwigs:MPP2_expt93.mm9.bw\n",
      "-- link_bigwigs:HSPC_expt93.mm9.bw\n",
      "-- link_bigwigs:S2_expts_13-35.mm9.bw\n",
      "-- link_bigwigs:S1_expts_10-13-35.mm9.bw\n",
      "-- link_bigwigs:S0_CD71_medium_expts_10-13-35.mm9.bw\n",
      "-- link_bigwigs:MPP3_expt93.mm9.bw\n",
      "-- link_bigwigs:LMPP_expt93.mm9.bw\n",
      "-- link_bigwigs:Lin_expt35.mm9.bw\n",
      "-- link_bigwigs:S3_expts_13-35.mm9.bw\n",
      ".  bigwig_hub_file:Making hub (mm9)\n",
      ".  bigwig_hub_file:Making hub (mm10)\n"
     ]
    }
   ],
   "source": [
    "%doit -n 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
